src/passes/Asyncify.cpp

```diff
@@ -681,7 +681,7 @@ class ModuleAnalyzer {
                                    !info.inRemoveList;
                           },
                           [verbose](Info& info, Function* reason) {
-                            if (verbose && !info.canChangeState) {
+                            if (verbose) {
                               std::cout << "[asyncify] " << info.name
                                         << " can change the state due to "
                                         << reason->name << "\n";
```

Comment:
Reading the logic in the `propagateBack` function, I suspect this is only called once for the first `reason` anyhow, because of the check for `!hasProperty`:

https://github.com/WebAssembly/binaryen/blob/b01246f3a9fd11c0849355794d88cb53da3b8258/src/ir/module-utils.h#L414-L415

That is once we add the property, we never add it again.

Maybe I've misunderstood the code though?

Reply:
I admit I haven't tested this, but that's because it's a very simple change...

Yes it only propagates once per reason, but it then calls this function for every caller of reason. We want it to log for every reason-caller pair. Currently it won't log if the caller has already been marked, so removing that check means all pairs will be logged.

Reply:
You could be right, but to me it looks like once one caller provides a reason, the target has the property, and then no other caller will be called. So only one pair is printed. But again, I may be misreading this. I'd recommend adding a test to verify things.

I do suspect you are right because by my reading of things, that check was unneeded code before anyhow, which seems odd...

Reply:
I can't see an existing test for the `asyncify-verbose` option here, and I'm not sure how the test framework here works.

Emscripten does have [a test](https://github.com/emscripten-core/emscripten/blob/f07067df82dd26f6c8fa58dcf97bac0da0d9259b/test/test_other.py#L10212), but it doesn't check for multiple logs of one function. I could make a PR to update that test in the future. Which doesn't mean it wouldn't be better to have a test here, but I'm really not sure how to do that sorry.

Reply:
> You could be right, but to me it looks like once one caller provides a reason, the target has the property, and then no other caller will be called. So only one pair is printed. But again, I may be misreading this. I'd recommend adding a test to verify things.

No, I think you're right. I didn't really fully understand what it was doing, but this line is the `hasProperty` function:
https://github.com/WebAssembly/binaryen/blob/b01246f3a9fd11c0849355794d88cb53da3b8258/src/passes/Asyncify.cpp#L678

I thought about whether it would be enough to just change it to `[](const Info& info) { return 0; }`, but I think that could cause infinite loops.

To get the logs I want, I think the logic of `propagateBack` would need to be reversed. Rather than checking whether the caller has been visited before, it would need to check whether the `reason` func has been visited...

Reply:
I think I've got it!

If we do the logging *in* the `hasProperty` function then we can log for ever caller-callee pair, while not changing how the propagation walks the tree. We'd just need to change the signature of `hasProperty` to add the `reason` func. But `propagateBack` is only called in two places, so that's not onerous.

Reply:
Okay, that's done. It's a bigger change than I thought this would require, but I think it's correct while still being relatively unobtrusive to the rest of the codebase. PTAL. 😃